src/sync.rs

```diff
@@ -208,6 +208,13 @@ trait EvalContextExtPriv<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
 // situations.
 impl<'mir, 'tcx: 'mir> EvalContextExt<'mir, 'tcx> for crate::MiriEvalContext<'mir, 'tcx> {}
 pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx> {
+    #[inline]
+    /// Peek the id of the next mutex
+    fn mutex_next_id(&self) -> MutexId {
```

Comment:
> This seems rather fragile... it only exists for these strange new orderings in the mutex implementation, right? I hope we can avoid this.^^

Unfortunately this isn't about the orderings but RMW. We can't conditionally call `mutex_create()` and get a new ID on a compare exchange success (i.e. the current value is 0). We must know the potential new ID beforehand.